go/vrf/ecvrf_p256_sha256_swu.go

```diff
@@ -0,0 +1,187 @@
+	x, y, _, _ = aux.hashToCurve(pub, alpha)
+	return
+}
+
+// hashToCurveSimplifiedSWU
+// Input:
+//    suite_string - a single octet specifying ECVRF ciphersuite.
+//    pub - public key, an EC point
+//    alpha_string - value to be hashed, an octet string
+// Output:
+//    Hx, Hy - hashed value, a finite EC point in G
+//
+// https://tools.ietf.org/html/draft-irtf-cfrg-vrf-06#section-5.4.1.3
+func (aux *p256SHA256SWUAux) hashToCurve(pub *PublicKey, alpha []byte) (hx, hy, t, w *big.Int) {
+	// Fixed options:
+	// a and b, constants for the Weierstrass form elliptic curve
+	// equation y^2 = x^3 + ax +b for the curve E
+	p := aux.params.ec.Params().P
+	a := big.NewInt(-3) // A Curve represents a short-form Weierstrass curve with a=-3.
+	b := aux.params.ec.Params().B
+	one := big.NewInt(1)
+	two := big.NewInt(2)
+
+	// 1.   PK_string = EC2OSP(Y)
+	// 2.   one_string = 0x01 = I2OSP(1, 1), a single octet with value 1
+	// 3.   t_string = Hash(suite_string || one_string || PK_string || alpha_string)
+	th := aux.params.hash.New()
+	th.Write([]byte{aux.params.suite, 0x01})
+	th.Write(aux.PointToString(pub.X, pub.Y))
+	th.Write(alpha)
+
+	// 4.   t = string_to_int(t_string) mod p
+	t = new(big.Int).SetBytes(th.Sum(nil))
+	t.Mod(t, p)
+
+	// 5.   r = -(t^2) mod p
+	r := new(big.Int).Mul(t, t)
+	r.Neg(r)
+	r.Mod(r, p)
+
+	// 6.   d = (r^2 + r) mod p   (d is t^4-t^2 mod p)
+	d := new(big.Int).Mul(r, r)
+	d.Add(d, r)
+	d.Mod(d, p)
+
+	// 7.   If d = 0 then d_inverse = 0; else d_inverse = 1/d mod p
+	//      (as long as Hash is secure, the case of d = 0 is an utterly
+	//      improbably occurrence;
+	//      the two cases can be combined into one by computing d_inverse = d^(p-2) mod p)
+	dI := new(big.Int).Exp(d, new(big.Int).Sub(p, two), p)
+
+	// 8.   x = ((-b/a) * (1 + d_inverse)) mod p
+	//      c = -b/a
+	c := new(big.Int).Mul(new(big.Int).Neg(b), new(big.Int).ModInverse(a, p))
+	c.Mod(c, p)
+	//      x = (c * (1 + d_inverse)) mod p
+	x := new(big.Int).Mul(c, new(big.Int).Add(one, dI))
+	x.Mod(x, p)
+
+	// 9.   w = (x^3 + a*x + b) mod p
+	//      (this step evaluates the curve equation)
+	w = new(big.Int).Mul(x, x)
+	w.Mul(w, x)
+	w.Add(w, new(big.Int).Mul(a, x))
+	w.Add(w, b)
+	w.Mod(w, p)
+
+	// 10.  Let e equal the Legendre symbol of w and p (see note below on how to compute e)
+	// 11.  If e is equal to 0 or 1 then final_x = x; else final_x = r * x mod p
+	//	  (final_x is the x-coordinate of the output; see note below on how to compute it)
+	//
+	//   In order to make this algorithm run in time that is (almost)
+	//   independent of the input (so-called "constant-time"), implementers
+	//   should pay particular attention to Steps 10 and 11 above.  These
+	//   steps can be implemented using the following approach:
+	//
+	//   If arithmetic and CMOV are implemented in constant time, then steps 9
+	//   and 10 above can be implemented as follows:
+	//
+	//      e = (w ^ ((p-1)/2))+1 mod p
+	//      (At this point, e is 0, 1, or 2, as an integer.)
```

Comment:
It think the code look correct according to the IETF document, but could there be an error in the document:

If `e = {0, 1, 2}`, then `b = {0, 1, 1}`. However, 11) says that if e is {0, 1} then we should use `x`. Otherwise `r*x`.
The cmov selects `x` when b = 1 (hence when e is 1 or 2), and `r*x` when b = 0. This seems to be an error?

Reply:
That is a good point. @reyzin - should the spec be modified? 

current text:
`11.  If e is equal to 0 or 1 then final_x = x; else final_x = r * x mod p`
observed behavior:
`11.  If e is equal to 1 or 2 then final_x = x; else final_x = r * x mod p`

Reply:
I take that back, upon closer inspection, [revision 07](https://tools.ietf.org/html/draft-irtf-cfrg-vrf-07#section-5.5) of the spec, this entire algorithm has been replaced with https://tools.ietf.org/html/draft-irtf-cfrg-hash-to-curve-07

Reply:
Ok, that also solves this :-). I noticed there are a few more smaller changes in the new version (mostly some 0 bytes added to the hashes).